pkg/services/annotations/annotationsimpl/loki/historian_store.go

```diff
@@ -48,13 +70,215 @@ func NewLokiHistorianStore(cfg setting.UnifiedAlertingStateHistorySettings, ft f
+			DashboardUID: &entry.DashboardUID,
+			PanelID:      entry.PanelID,
+			NewState:     entry.Current,
+			PrevState:    entry.Previous,
+			Time:         sample.T.UnixMilli(),
+			Text:         annotationText,
+			Data:         annotationData,
+		})
+	}
+
+	return items
 }
 
 func (r *LokiHistorianStore) GetTags(ctx context.Context, query *annotations.TagsQuery) (annotations.FindTagsResult, error) {
 	return annotations.FindTagsResult{}, nil
 }
 
+// util
+
+func getRule(ctx context.Context, sql db.DB, orgID int64, ruleID int64) (*ngmodels.AlertRule, error) {
+	rule := &ngmodels.AlertRule{OrgID: orgID, ID: ruleID}
+	err := sql.WithDbSession(ctx, func(sess *db.Session) error {
+		exists, err := sess.Get(rule)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return ErrMissingRule
+		}
+		return nil
+	})
+
+	return rule, err
+}
+
+func hasAccess(entry historian.LokiEntry, resources accesscontrol.AccessResources) bool {
+	if resources.CanAccessOrgAnnotations && entry.DashboardUID != "" {
+		return false
+	}
+
+	if resources.CanAccessDashAnnotations {
+		_, canAccess := resources.Dashboards[entry.DashboardUID]
+		if !canAccess {
+			return false
+		}
+	}
+
+	return true
+}
+
+type number interface {
+	constraints.Integer | constraints.Float
+}
+
+// numericMap converts a simplejson map[string]any to a map[string]N, where N is numeric (int or float).
+func numericMap[N number](j *simplejson.Json) (map[string]N, error) {
+	m, err := j.Map()
+	if err != nil {
+		return nil, err
+	}
+
+	values := make(map[string]N)
+	for k, v := range m {
+		a, ok := (v).(json.Number)
+		if !ok {
+			return nil, fmt.Errorf("unexpected value type %T", v)
+		}
+
+		f, err := a.Float64()
+		if err != nil {
+			return nil, err
+		}
+
+		values[k] = N(f)
+	}
+
+	return values, nil
+}
+
+func buildTransition(entry historian.LokiEntry) (*state.StateTransition, error) {
+	curState, curStateReason, err := state.ParseFormattedState(entry.Current)
+	if err != nil {
+		return nil, fmt.Errorf("parsing current state: %w", err)
+	}
+
+	prevState, prevReason, err := state.ParseFormattedState(entry.Previous)
+	if err != nil {
+		return nil, fmt.Errorf("parsing previous state: %w", err)
+	}
+
+	v, err := numericMap[float64](entry.Values)
+	if err != nil {
+		return nil, fmt.Errorf("parsing entry values: %w", err)
+	}
+
+	return &state.StateTransition{
+		State: &state.State{
+			State:       curState,
+			StateReason: curStateReason,
+			Values:      v,
+			Labels:      entry.InstanceLabels,
+		},
+		PreviousState:       prevState,
+		PreviousStateReason: prevReason,
+	}, nil
+}
+func buildHistoryQuery(ctx context.Context, query *annotations.ItemQuery, dashboards map[string]int64, ruleUID string) ngmodels.HistoryQuery {
+	historyQuery := ngmodels.HistoryQuery{
+		OrgID:        query.OrgID,
+		DashboardUID: query.DashboardUID,
+		PanelID:      query.PanelID,
+		RuleUID:      ruleUID,
+	}
+
+	if historyQuery.DashboardUID == "" && query.DashboardID == 0 {
```

Comment:
Is supposed to be
```suggestion
	if historyQuery.DashboardUID == "" && query.DashboardID != 0 {
```

If so, probably add a test to cover as well.